java.security.PrivilegedActionException: java.security.PrivilegedActionException: java.io.IOException: Cannot delete 'C:\Users\reineke\IdeaProjects\Riddle\out\main.synctex(busy)'

### Type of JetBrains IDE (IntelliJ, PyCharm, etc.) and version
IntelliJ IDEA 2021.3.1 (build IU-213.6461.79)

### Operating System 
Windows 10 10.0 (amd64)

### TeXiFy IDEA version
0.7.13

### Description



### Stacktrace
```
java.security.PrivilegedActionException: java.security.PrivilegedActionException: java.io.IOException: Cannot delete 'C:\Users\reineke\IdeaProjects\Riddle\out\main.synctex(busy)'
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:891)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:757)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:447)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:818)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:446)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:805)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:498)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.security.PrivilegedActionException: java.io.IOException: Cannot delete 'C:\Users\reineke\IdeaProjects\Riddle\out\main.synctex(busy)'
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	... 17 more
Caused by: java.io.IOException: Cannot delete 'C:\Users\reineke\IdeaProjects\Riddle\out\main.synctex(busy)'
	at com.intellij.openapi.vfs.impl.local.LocalFileSystemBase.deleteFile(LocalFileSystemBase.java:373)
	at com.intellij.openapi.vfs.newvfs.persistent.PersistentFSImpl.deleteFile(PersistentFSImpl.java:532)
	at com.intellij.openapi.vfs.newvfs.impl.VirtualFileSystemEntry.delete(VirtualFileSystemEntry.java:261)
	at nl.hannahsten.texifyidea.action.DeleteGeneratedFiles$deleteFiles$5.invoke(DeleteGeneratedFiles.kt:73)
	at nl.hannahsten.texifyidea.action.DeleteGeneratedFiles$deleteFiles$5.invoke(DeleteGeneratedFiles.kt:71)
	at nl.hannahsten.texifyidea.util.GeneralKt.runWriteAction$lambda-0(General.kt:44)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:947)
	at nl.hannahsten.texifyidea.util.GeneralKt.runWriteAction(General.kt:44)
	at nl.hannahsten.texifyidea.action.DeleteGeneratedFiles.deleteFiles(DeleteGeneratedFiles.kt:71)
	at nl.hannahsten.texifyidea.action.DeleteGeneratedFiles.actionPerformed(DeleteGeneratedFiles.kt:23)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.lambda$performActionDumbAwareWithCallbacks$4(ActionUtil.java:244)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:265)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAwareWithCallbacks(ActionUtil.java:244)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.lambda$performAction$0(ActionMenuItem.java:277)
	at com.intellij.openapi.wm.impl.FocusManagerImpl.runOnOwnContext(FocusManagerImpl.java:236)
	at com.intellij.openapi.wm.impl.IdeFocusManagerImpl.runOnOwnContext(IdeFocusManagerImpl.java:67)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.performAction(ActionMenuItem.java:269)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem$ActionTransmitter.actionPerformed(ActionMenuItem.java:284)
	at java.desktop/javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:1967)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.lambda$fireActionPerformed$2(ActionMenuItem.java:111)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:94)
	at com.intellij.openapi.actionSystem.impl.ActionMenuItem.fireActionPerformed(ActionMenuItem.java:111)
	at com.intellij.ui.plaf.beg.BegMenuItemUI.doClick(BegMenuItemUI.java:541)
	at com.intellij.ui.plaf.beg.BegMenuItemUI$MyMouseInputHandler.mouseReleased(BegMenuItemUI.java:571)
	at java.desktop/java.awt.Component.processMouseEvent(Component.java:6654)
	at java.desktop/javax.swing.JComponent.processMouseEvent(JComponent.java:3345)
	at java.desktop/java.awt.Component.processEvent(Component.java:6419)
	at java.desktop/java.awt.Container.processEvent(Container.java:2263)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5029)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2321)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4861)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4918)
	at java.desktop/java.awt.LightweightDispatcher.processMouse

--- a/src/nl/hannahsten/texifyidea/action/DeleteGeneratedFiles.kt
+++ b/src/nl/hannahsten/texifyidea/action/DeleteGeneratedFiles.kt
@@ -77,7 +77,8 @@ class DeleteGeneratedFiles : AnAction() {
                 path?.children?.forEach {
                     try {
                         it.delete(this)
-                    } catch (e: IOException) {
+                    }
+                    catch (e: IOException) {
                         notDeleted.add(it)
                     }
                 }

--- a/src/nl/hannahsten/texifyidea/action/DeleteGeneratedFiles.kt
+++ b/src/nl/hannahsten/texifyidea/action/DeleteGeneratedFiles.kt
@@ -7,10 +7,12 @@ import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.ui.Messages
 import com.intellij.openapi.ui.showOkCancelDialog
 import com.intellij.openapi.vfs.LocalFileSystem
+import com.intellij.openapi.vfs.VirtualFile
 import nl.hannahsten.texifyidea.util.getLatexRunConfigurations
 import nl.hannahsten.texifyidea.util.magic.FileMagic
 import nl.hannahsten.texifyidea.util.runWriteAction
 import java.io.File
+import java.io.IOException
 import java.security.PrivilegedActionException
 
 /**
@@ -67,11 +69,23 @@ class DeleteGeneratedFiles : AnAction() {
             path.walk().maxDepth(1).forEach { it.delete() }
         }
 
+        val notDeleted = mutableListOf<VirtualFile>()
+
         // Custom out/aux dirs
         runWriteAction {
             for (path in customOutput) {
-                path?.children?.forEach { it.delete(this) }
+                path?.children?.forEach {
+                    try {
+                        it.delete(this)
+                    } catch (e: IOException) {
+                        notDeleted.add(it)
+                    }
+                }
             }
+        }
+
+        if (notDeleted.isNotEmpty()) {
+            Notification("LaTeX", "Could not delete some files", "The following files need to be deleted manually: ${notDeleted.map { it.path }}", NotificationType.WARNING).notify(project)
         }
 
         // Generated minted files

--- a/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexEnterInCommentHandler.kt
+++ b/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexEnterInCommentHandler.kt
@@ -52,7 +52,8 @@ class LatexEnterInCommentHandler : EnterHandlerDelegateAdapter() {
         if (!currentCaret.isValid) return
         try {
             PsiUtilCore.ensureValid(file)
-        } catch (e: PsiInvalidElementAccessException) {
+        }
+        catch (e: PsiInvalidElementAccessException) {
             return
         }
 

--- a/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexEnterInCommentHandler.kt
+++ b/src/nl/hannahsten/texifyidea/editor/typedhandlers/LatexEnterInCommentHandler.kt
@@ -9,6 +9,8 @@ import com.intellij.openapi.util.Ref
 import com.intellij.psi.PsiComment
 import com.intellij.psi.PsiDocumentManager
 import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiInvalidElementAccessException
+import com.intellij.psi.util.PsiUtilCore
 import nl.hannahsten.texifyidea.psi.LatexMagicComment
 import nl.hannahsten.texifyidea.util.files.document
 import nl.hannahsten.texifyidea.util.get
@@ -45,7 +47,16 @@ class LatexEnterInCommentHandler : EnterHandlerDelegateAdapter() {
         // Check location of the % on the previous line, especially the number of spaces before and after it
         // because that's what we want to have on the current line as well
         // (note the cursor is on the second line, after the enter has been processed)
-        val previousLineNumber = (editor.caretModel.currentCaret.logicalPosition.line - 1).coerceAtLeast(0)
+        val currentCaret = editor.caretModel.currentCaret
+
+        if (!currentCaret.isValid) return
+        try {
+            PsiUtilCore.ensureValid(file)
+        } catch (e: PsiInvalidElementAccessException) {
+            return
+        }
+
+        val previousLineNumber = (currentCaret.logicalPosition.line - 1).coerceAtLeast(0)
         val lineStart = editor.document.getLineStartOffset(previousLineNumber)
         val lineEnd = editor.document.getLineEndOffset(previousLineNumber)
         val previousLine = editor.document[IntRange(lineStart, lineEnd)]

--- a/src/nl/hannahsten/texifyidea/util/Strings.kt
+++ b/src/nl/hannahsten/texifyidea/util/Strings.kt
@@ -9,9 +9,9 @@ import kotlin.math.max
 import kotlin.math.min
 
 /**
- * Capitalises the first character of the string.
+ * Capitalises the first character of the string, if present.
  */
-fun String.capitalizeFirst(): String = this[0].toUpperCase() + substring(1, length)
+fun String.capitalizeFirst(): String = if (this.isEmpty()) this else this[0].toUpperCase() + substring(1, length)
 
 /**
  * Converts the string to camel case.
@@ -105,7 +105,7 @@ fun List<String>.removeIndents(): List<String> {
 
     val list = ArrayList<String>(size)
     val (maxIndent, _) = asSequence()
-        .filter { !it.isBlank() }
+        .filter { it.isNotBlank() }
         .map { Pair(it.getIndent().length, it) }
         .minByOrNull { it.first } ?: return this
 
@@ -167,7 +167,7 @@ fun String.formatAsFilePath(): String {
         .toLowerCase()
 
     // If there are no valid characters left, use a default name.
-    return if (formatted.isEmpty()) "myfile" else formatted
+    return formatted.ifEmpty { "myfile" }
 }
 
 /**
@@ -189,7 +189,7 @@ fun String.splitWhitespace() = split(Regex("\\s+"))
  *
  * @return The string with HTML tags removed.
  *
- * @see [Magic.Pattern.htmlTag]
+ * @see [PatternMagic.htmlTag]
  */
 fun String.removeHtmlTags() = this.replace(PatternMagic.htmlTag.toRegex(), "")
 

